### PR TITLE
Issue #143: 7.x-2.x: Optimize atomium_get_settings()

### LINKED ISSUE #143: Optimize atomium_get_settings()
Since I opened #129 as a generic "meta" issue, I am opening this one as a practical issue that will actually result in a PR.

atomium_get_settings() called 20 times on my local test site, and each time it calls _atomium_array_flatten().

### PR DESCRIPTION
**Pull request by donquixote**
Fixes #143.

This is a backport for 7.x-2.x based on the first two commits in https://github.com/ec-europa/atomium/pull/147.

If my profiling is correct:
- The first two commits of the other PR each bring a clear performance improvement.
- The 3rd commit of the other PR provides only a negligible improvement, but has a risk of BC breakage in case of setting keys containing '.' character (which is used as glue in _atomium_array_flatten()). This is why this last commit is not included here.

While the other PR was marked "WIP", with this one I am serious :)

### CODE REVIEW

**Comment by donquixote** [comment]
Rebased on 888d4be9d4962a3e36a4a6e88cb9d7db4707fccb, which is the common ancestor of 7.x-2.x and 7.x-3.x.

This should make it possible to merge this into both branches without conflict.
(but start with 7.x-2.x)

File: atomium/includes/common.inc [hunk 1]
```diff
@@ -667,33 +667,125 @@ function _atomium_array_flatten(array $array, array &$storage, $parentKey = '')
...
-    );
+/**
+ * @param string $theme_key
+ * @param bool $base_themes
+ * @param callable[] $value_callbacks
+ *
+ * @return mixed
+ */
+function _atomium_get_settings_flat_filtered($theme_key, $base_themes = TRUE, $value_callbacks = ['trim']) {
+  static $settings_flat_filtered_by_cid = [];
```
Comment (imanoleguskiza) [inline atomium/includes/common.inc:685]: Why not use drupal_static?

File: atomium/includes/common.inc [hunk 2]
```diff
@@ -667,33 +667,125 @@ function _atomium_array_flatten(array $array, array &$storage, $parentKey = '')
...
+    ? $settings_flat_filtered[$setting_keys]
+    : [];
+}
 
-  if ($theme_settings = atomium_get_theme_info($theme_key, 'settings', $base_themes)) {
-    _atomium_array_flatten(
-      $theme_settings,
-      $settings
-    );
+/**
```
Comment (imanoleguskiza) [inline atomium/includes/common.inc:677]: Function needs a description

File: atomium/includes/common.inc [hunk 3]
```diff
@@ -667,33 +667,125 @@ function _atomium_array_flatten(array $array, array &$storage, $parentKey = '')
...
+  }
+
+  if (NULL !== $cid) {
+    $settings_flat_filtered_by_cid[$cid] = $theme_settings_flat;
+  }
+
+  return $theme_settings_flat;
+}
+
+/**
```
Comment (imanoleguskiza) [inline atomium/includes/common.inc:764]: Function needs a description

**Review by imanoleguskiza — CHANGES_REQUESTED** [review]
Could we have information on how much performance gain this PR would bring? It seems to me we are trading a bit of performance for much more simpler code.

File: atomium/includes/common.inc (hunk 2 above)
Comment (drupol) [reply atomium/includes/common.inc:677]: Fixed in #201

File: atomium/includes/common.inc (hunk 3 above)
Comment (drupol) [reply atomium/includes/common.inc:764]: Fixed in #201

**Comment by drupol** [comment]
Closed for #201

File: atomium/includes/common.inc (hunk 1 above)
Comment (drupol) [reply atomium/includes/common.inc:685]: Please could you address your remarks in #201 ?

**Pull request closed without merging**